proof_ipa.go

```diff
@@ -138,50 +138,27 @@ func getProofElementsFromTree(preroot, postroot VerkleNode, keys [][]byte, resol
 		return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting pre-state proof data: %w", err)
 	}
 
-	// List of points and vectors and indices used to generate the IPA proof
-	// If only a pre-state root is available, it's a simple copy, but if both
-	// pre- and post-state root are present, it's merging the two lists.
-	var (
-		proof_cis []*Point
-		proof_fs  [][]Fr
-		proof_zs  []uint8
-		proof_ys  []*Fr
-		postvals  = make([][]byte, len(keys))
-	)
-	for i := range pe.Cis {
-		proof_cis = append(proof_cis, pe.Cis[i])
-		proof_fs = append(proof_fs, pe.Fis[i])
-		proof_ys = append(proof_ys, pe.Yis[i])
-		proof_zs = append(proof_zs, pe.Zis[i])
-	}
-
 	// if a post-state tree is present, merge its proof elements with
 	// those of the pre-state tree, so that they can be proved together.
+	postvals := make([][]byte, len(keys))
 	if postroot != nil {
-		pe_post, _, _, err := GetCommitmentsForMultiproof(postroot, keys, resolver)
-		if err != nil {
-			return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting post-state proof data: %w", err)
-		}
-
-		for i := range pe_post.Cis {
-			proof_cis = append(proof_cis, pe_post.Cis[i])
-			proof_fs = append(proof_fs, pe_post.Fis[i])
-			proof_ys = append(proof_ys, pe_post.Yis[i])
-			proof_zs = append(proof_zs, pe_post.Zis[i])
-		}
-
+		// keys were sorted already in the above GetcommitmentsForMultiproof.
 		// Set the post values, if they are untouched, leave them `nil`
-		for i, v := range pe.Vals {
-			if !bytes.Equal(v, pe_post.Vals[i]) {
-				postvals[i] = pe_post.Vals[i]
+		for i := range keys {
+			val, err := postroot.Get(keys[i], resolver)
+			if err != nil {
+				return nil, nil, nil, nil, nil, nil, nil, nil, fmt.Errorf("error getting post-state value for key %x: %w", keys[i], err)
+			}
+			if !bytes.Equal(pe.Vals[i], val) {
+				postvals[i] = val
 			}
```

Comment:
This is something that can be optimizable if we allow the client to do the work.

The client (geth), after the EVM block is executed, has full knowledge of which are all the post-state values, so there's no need to do this L147-L153 here.

This current implementation isn't that bad, since actually `postroot` is already hot in memory (i.e: all nodes resolved) since Geth already updated the tree. So technically, it shouldn't need to resolve any node thus should be quite fast.

In any case, if that's convincing then maybe this is good enough and can avoid some coupling. But, would be worth double-checking.

Reply:
yeah let's do it in the client, using the access witness if possible.